File: benchmarks/pandas/bench_extensions.py
"""Benchmark: pd.api.extensions — ExtensionDtype / ExtensionArray subclassing and
accessor registration.

Mirrors tsb's extensions benchmark:
    - ExtensionDtype subclass construction
    - ExtensionArray subclass instantiation, getitem, slice, dtype access
    - register_extension_dtype()           → tsb registerExtensionDtype()
    - register_series_accessor()           → tsb registerSeriesAccessor()
    - register_dataframe_accessor()        → tsb registerDataFrameAccessor()
    - Accessor registry introspection via hasattr
"""
import json
import time
import numpy as np
import pandas as pd
import pandas.api.extensions as pd_ext

WARMUP = 5
ITERATIONS = 200


@pd_ext.register_extension_dtype
class TagDtype(pd_ext.ExtensionDtype):
    name = "tag"
    type = object
    kind = "O"

    @classmethod
    def construct_array_type(cls):
        return TagArray

    @classmethod
    def construct_from_string(cls, string):
        if string == "tag":
            return cls()
        raise TypeError(f"Cannot construct a 'TagDtype' from '{string}'")


class TagArray(pd_ext.ExtensionArray):
    def __init__(self, data):
        self._data = np.asarray(data, dtype=object)

    @classmethod
    def _from_sequence(cls, scalars, *, dtype=None, copy=False):
        return cls(scalars)

    @classmethod
    def _from_factorized(cls, values, original):
        return cls(values)

    def __getitem__(self, key):
        return self._data[key]

    def __setitem__(self, key, value):
        self._data[key] = value

    def __len__(self):
        return len(self._data)

    @property
    def dtype(self):
        return TagDtype()

    @property
    def nbytes(self):
        return self._data.nbytes

    def isna(self):
        return np.array([v is None for v in self._data])

    def take(self, indices, *, allow_fill=False, fill_value=None):
        return type(self)(self._data.take(indices))

    def copy(self):
        return type(self)(self._data.copy())

    @classmethod
    def _concat_same_type(cls, to_concat):
        return cls(np.concatenate([a._data for a in to_concat]))


@pd_ext.register_series_accessor("geo_bench")
class GeoAccessor:
    def __init__(self, obj):
        self._obj = obj

    def distance(self):
        return 0


@pd_ext.register_dataframe_accessor("geo_bench")
class GeoDataFrameAccessor:
    def __init__(self, obj):
        self._obj = obj

    def distance(self):
        return 0


_TAGS = ["alpha", "beta", "gamma", "delta", "epsilon"]
_s = pd.Series(TagArray(_TAGS))
_df = pd.DataFrame({"a": [1, 2, 3]})


def _run():
    arr = TagArray(_TAGS)
    _len = len(arr)
    _item = arr[2]
    _sliced = arr[1:4]
    _dtype_name = arr.dtype.name
    _numeric = False

    _has_series = hasattr(_s, "geo_bench")
    _has_df = hasattr(_df, "geo_bench")

    return [_len, _item, _sliced, _dtype_name, _numeric, _has_series, _has_df]


for _ in range(WARMUP):
    _run()

start = time.perf_counter()
for _ in range(ITERATIONS):
    _run()
total_ms = (time.perf_counter() - start) * 1000

print(
    json.dumps(
        {
            "function": "extensions",
            "mean_ms": total_ms / ITERATIONS,
            "iterations": ITERATIONS,
            "total_ms": total_ms,
        }
    )
)

File: benchmarks/tsb/bench_extensions.ts
/**
 * Benchmark: pd.api.extensions — ExtensionDtype / ExtensionArray / accessor registration.
 *
 * Covers:
 *   - ExtensionDtype subclassing              → pandas `pandas.api.extensions.ExtensionDtype`
 *   - ExtensionArray subclassing              → pandas `pandas.api.extensions.ExtensionArray`
 *   - registerExtensionDtype()                → pandas `register_extension_dtype()`
 *   - constructExtensionDtypeFromString()     → pandas dtype string resolution
 *   - registerSeriesAccessor()                → pandas `register_series_accessor()`
 *   - registerDataFrameAccessor()             → pandas `register_dataframe_accessor()`
 *   - getRegisteredAccessors()                → accessor registry lookup
 *
 * Outputs JSON: {"function": "extensions", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import {
  ExtensionDtype,
  ExtensionArray,
  registerExtensionDtype,
  constructExtensionDtypeFromString,
  registerSeriesAccessor,
  registerDataFrameAccessor,
  getRegisteredAccessors,
} from "../../src/index.ts";

const WARMUP = 5;
const ITERATIONS = 200;

class TagDtype extends ExtensionDtype {
  override get name(): string {
    return "tag";
  }
  override get type(): abstract new (...args: readonly unknown[]) => unknown {
    return String as unknown as abstract new (...args: readonly unknown[]) => unknown;
  }
  override get kind(): string {
    return "O";
  }
  override get isNumeric(): boolean {
    return false;
  }
  static override construct_from_string(dtype: string): TagDtype | null {
    return dtype === "tag" ? new TagDtype() : null;
  }
}

class TagArray extends ExtensionArray {
  private readonly _data: readonly string[];
  constructor(data: readonly string[]) {
    super();
    this._data = data;
  }
  override get dtype(): TagDtype {
    return new TagDtype();
  }
  override get length(): number {
    return this._data.length;
  }
  override getItem(i: number): string | null {
    const idx = i < 0 ? this._data.length + i : i;
    return this._data[idx] ?? null;
  }
  override slice(start: number, stop: number): TagArray {
    return new TagArray(this._data.slice(start, stop));
  }
}

class GeoAccessor {
  constructor(_obj: unknown) {}
  distance(): number {
    return 0;
  }
}

// Register once — idempotent for repeated benchmark runs
registerExtensionDtype(TagDtype as unknown as { new (): ExtensionDtype } & typeof ExtensionDtype);
registerSeriesAccessor("geo_bench", GeoAccessor);
registerDataFrameAccessor("geo_bench", GeoAccessor);

function run(): void {
  const dt = constructExtensionDtypeFromString("tag");
  const _name = dt?.name;

  const arr = new TagArray(["alpha", "beta", "gamma", "delta", "epsilon"]);
  const _len = arr.length;
  const _item = arr.getItem(2);
  const _neg = arr.getItem(-1);
  const _sliced = arr.slice(1, 4);
  const _dtype = arr.dtype.name;
  const _numeric = arr.dtype.isNumeric;

  const seriesMap = getRegisteredAccessors("series");
  const _hasSeries = seriesMap.has("geo_bench");
  const dfMap = getRegisteredAccessors("dataframe");
  const _hasDf = dfMap.has("geo_bench");
  const idxMap = getRegisteredAccessors("index");
  const _idxSize = idxMap.size;

  void [_name, _len, _item, _neg, _sliced, _dtype, _numeric, _hasSeries, _hasDf, _idxSize];
}

for (let i = 0; i < WARMUP; i++) run();

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) run();
const total_ms = performance.now() - start;

console.log(
  JSON.stringify({
    function: "extensions",
    mean_ms: total_ms / ITERATIONS,
    iterations: ITERATIONS,
    total_ms,
  }),
);
